src/history/hs.c

```diff
@@ -1236,121 +1236,186 @@ __hs_delete_key_from_pos(
+         */
+        if (btree_id != this_btree_id)
             break;
 
         /*
-         *  Keep track of the previous comparison. The history store is stored in order, so we can
-         *  avoid redundant comparisons. Previous btree ID isn't set, until data cursor is open.
+         * If we have already checked against this key, keep going to the next key. We only need to
+         * check the key once.
          */
-        if (data_cursor == NULL || (prev_btree_id != btree_id)) {
-            /*
-             * Check whether this btree-id exists in the metadata. We do that by finding what uri
-             * this btree belongs to. Using this URI, verify the history store key with the data
-             * store.
-             */
-            if (data_cursor != NULL) {
-                WT_ERR(data_cursor->close(data_cursor));
-                /* Setting data_cursor to null, to avoid double free */
-                data_cursor = NULL;
-            }
+        WT_ERR(__wt_compare(session, NULL, hs_key, prev_hs_key, &cmp));
+        if (cmp == 0)
+            continue;
+        WT_WITH_PAGE_INDEX(session, ret = __wt_row_search(cbt, hs_key, false, NULL, false, NULL));
+        WT_ERR(ret);
+
+        found = cbt->compare == 0;
+        WT_ERR(__cursor_reset(cbt));
+
+        if (!found)
             /*
-             * Using the btree-id find the metadata entry and extract the URI for this btree. Don't
-             * forget to free the copy of the URI returned.
-             *
-             * Re-purpose the previous-key buffer on error, safe because we're about to error out.
+             * Although WT_NOTFOUND makes sense, it is already in use when we reach the end of the
+             * table and callers use that for that purpose. If we find an inconsistency here,
+             * return a different error instead.
              */
-            __wt_free(session, uri_itr);
-            if ((ret = __wt_metadata_btree_id_to_uri(session, btree_id, &uri_itr)) != 0)
-                WT_ERR_MSG(session, ret,
-                  "Unable to find btree-id %" PRIu32
-                  " in the metadata file for the associated history store key %s",
-                  btree_id,
-                  __wt_buf_set_printable(session, hs_key->data, hs_key->size, prev_hs_key));
-
-            WT_ERR(__wt_open_cursor(session, uri_itr, NULL, NULL, &data_cursor));
-            F_SET(data_cursor, WT_CURSOR_RAW_OK);
-        } else {
-            WT_ERR(__wt_compare(session, NULL, hs_key, prev_hs_key, &cmp));
-            if (cmp == 0)
-                continue;
-        }
-        WT_ERR(__wt_buf_set(session, prev_hs_key, hs_key->data, hs_key->size));
-        prev_btree_id = btree_id;
+            WT_ERR_MSG(session, WT_ERROR,
+              "the associated history store key %s was not found in the data store %s",
+              __wt_buf_set_printable(session, hs_key->data, hs_key->size, prev_hs_key),
+              session->dhandle->name);
```

Comment:
Isn't this a panic?

Reply:
Fixed. The original had `WT_NOTFOUND`. Thanks. This should be a panic.